src/quickfix.c

```diff
@@ -2175,6 +2176,8 @@ qf_add_entry(
 {
     qfline_T	*qfp;
     qfline_T	**lastp;	// pointer to qf_last or NULL
+    char_u	*fullname = fix_fname(fname);
```

Comment:
Both filenames refer to the same file.  So I am not sure whether this is a significant issue that should be addressed.

Reply:
I believe my use case(LSP + a build infrastructure that makes hard links) is not very common. I just saw it and try to see if I can make it better. This issue makes me seeing alot of duplicated entries when looking on references of a function. If they should be treated as same file, then the entries shouldn't be duplicated. But this also depends on which component we think should handle this case:

1. vim: This is what my patch tries to deals with.
2. the plugin or LSP: LSP shouldn't send two names in the responses if they are already the same files.
3. Users:  Hard links are OS specific and just make a lot of mess in general. Users shouldn't expect editors/LSPs/plugins to know well on how files are linked.

As I see that vim only shows one filename when the file's hard link is opened(instead of showing another name). I believe this implementation was thought(and tested) more extensively and the better ways to address this issue might be 2. or 3. If that's the case then we can just close this PR.